### ISSUE TITLE
[Fix][3.3.5] Creature's Pet always use run instead of walk when they follow the owner

### ISSUE THREAD

**Issue by kldxzzhc2** [TrinityCore/TrinityCore#12817]
db : up-to-date
core : 91053d557ca89c4b0c455366afae258835bd25f8

When a creature (A) summon another creature (B) as itself's pet, the pet (B) is always run to follow its owner (A). But most time, it should use walk. 

For example, the creature 3265 (which movementtype is 1) cast spell 6479 to summon creature 3939 as a pet. In game, you can see the the pet run a few steps, stop, run, stop. Not walk follow the owner.

TargetedMovementGenerator.cpp Line 111 change it to
```c++
	Unit * u = owner->GetCharmerOrOwner();
	if (u && 
		i_target->GetGUID() == u->GetGUID() && 
		u->GetTypeId() == TYPEID_UNIT &&
		i_target->ToCreature()->GetDefaultMovementType() == RANDOM_MOTION_TYPE) {
		init.SetWalk(true);
	} else {
		init.SetWalk(((D*)this)->EnableWalking());
	}
```

**Comment by tkrokli** [comment]
There is no need to change `TargetedMovementGenerator.cpp` at all. Reason: the [[Razormane Wolf]](http://wotlk.openwow.com/npc=3939)(3939) tries to follow its owner, [[Razormane Hunter]](http://wotlk.openwow.com/npc=3265)(3265), but NPC 3265 does not have a set of waypoints to follow. Razormane Hunter has only got random movement within a limited area, *just enough* to trigger its pet to run around its owner to place itself on its owner's left side, to be ready for a walk along a waypoint route (but there is no route). These NPCs used to patrol in wide loops or back and forth along a path of anything between 25 and 100 meters in-game, but that path is no longer there.

To test my theory, I went to The Barrens (`.tele Barrens`), used `.cast 6479` to spawn my own Razormane Wolf (using an undead warlock) and started to walk instead of running. Sure enough, the wolf pet will follow its owner's speed, so the reason that the Razormane Wolves are running in short bursts instead of walking, is simply because their owners have lost their set of waypoints.

![wowscrnshot_081615_080358](https://cloud.githubusercontent.com/assets/4903600/9292244/a6bef0aa-43ed-11e5-98d2-d397fa444104.jpg)

Some of the Razormane Hunters + their wolves hardly move at all. Example: GUID 20309, the first entry in the database (to go directly to this NPC, just type `.go creature Razormane_Hunter`) and you will see that the Razormane hunter is not even moving its feet, only looking from side to side. Its pet only uses its default sniffing and scratching emotes, nothing else. (see image above for location)

**TL;DR:** If the Razormane Hunters (NPC 3265) get their waypoints back, their pets will walk normally.

**Comment by tkrokli** [comment]
This issue can be closed, source is working as intended (but waypoints are still missing).

**Comment by Killyana** [comment]
This issue is not fixed as this PR wasn't committed: https://github.com/TrinityCore/TrinityCore/pull/15749

**Comment by Killyana** [comment]
I discovered a weird behavior:
- .go c id 32711
- notice the pet is using the run speed to follow his master (it must use the same speed mode as the master)
- .mod faction 14 0 0  on 32711  (keep the gm mode on)
- Hit the npc 32711 with any spell wothout killing it (keep the gm mode on)
- Notice the pet now is using the walk speed

**Comment by Killyana** [comment]
I tested this changes, and it fixes the issue.
https://github.com/TrinityCore/TrinityCore/commit/d7a7a65d96d9aebb6164571dd51016eb29d25ac5

**Comment by tkrokli** [comment]
TrinityCore rev. b58a364192ee 2017-06-11 02:31:59 -0300 (3.3.5 branch) (Win64, Release, Static)
TDB 335.63 + updates up to and including `2017_06_10_00_world.sql`

My comment https://github.com/TrinityCore/TrinityCore/issues/12817#issuecomment-131494240 is no longer valid (and may have been highly incorrect for quite some time), the Razormane Wolf runs in short bursts alongside the walking Razormane Hunters (just to clear that misinformation away). Killyana is right.

**Comment by MrSmite** [comment]
Does this have any effect on player pets?

There is a similar bug (I think reported years ago) where if a Hunter puts their pet on stay, moves far away and then sets their mode to "walk", the pet will walk the entire distance. In reality, the pet should only inherit the owner's "walk" mode if they are within follow distance.

My speed fix which helped stop them from despawning did not address this issue. For whatever reason, when the player switches from run to walk, the pet is instantly put in the same mode regardless of how far away they are.

**Comment by Killyana** [comment]
No, you have to check this PR https://github.com/TrinityCore/TrinityCore/pull/15749 to find:
- Only sync speed with owner when near or slower
- Pets only walk when near their owner

**Comment by MrSmite** [comment]
> No, you have to check this PR #15749 to find:

Hmm, thanks. Hadn't noticed that one since it's been closed without merge (I rarely check closed issues).

**Comment by Tyrelis** [comment]
This is still and issue check warp huntress kula in dalaran has same issue.
NPC ID: http://www.wowhead.com/npc=32711/warp-huntress-kula

**Comment by Aokromes** [comment]
```cpp
From d7a7a65d96d9aebb6164571dd51016eb29d25ac5 Mon Sep 17 00:00:00 2001
From: gashole <gashole@users.noreply.github.com>
Date: Sat, 31 Oct 2015 11:18:23 -0700
Subject: [PATCH] Core/Movement: Creature pets can now walk

---
 src/server/game/Movement/Spline/MoveSplineInit.cpp | 17 ++++++-----------
 1 file changed, 6 insertions(+), 11 deletions(-)

diff --git a/src/server/game/Movement/Spline/MoveSplineInit.cpp b/src/server/game/Movement/Spline/MoveSplineInit.cpp
index 180ecf230799..87512bff7b1f 100644
--- a/src/server/game/Movement/Spline/MoveSplineInit.cpp
+++ b/src/server/game/Movement/Spline/MoveSplineInit.cpp
@@ -96,18 +96,13 @@ namespace Movement
         if (moveFlags & MOVEMENTFLAG_ROOT)
             moveFlags &= ~MOVEMENTFLAG_MASK_MOVING;
 
-        if (!args.HasVelocity)
-        {
-            // If spline is initialized with SetWalk method it only means we need to select
-            // walk move speed for it but not add walk flag to unit
-            uint32 moveFlagsForSpeed = moveFlags;
-            if (args.flags.walkmode)
-                moveFlagsForSpeed |= MOVEMENTFLAG_WALKING;
-            else
-                moveFlagsForSpeed &= ~MOVEMENTFLAG_WALKING;
+        if (args.flags.walkmode)
+            moveFlags |= MOVEMENTFLAG_WALKING;
+        else
+            moveFlags &= ~MOVEMENTFLAG_WALKING;
 
-            args.velocity = unit->GetSpeed(SelectSpeedType(moveFlagsForSpeed));
-        }
+        if (!args.HasVelocity)
+            args.velocity = unit->GetSpeed(SelectSpeedType(moveFlags));
 
         if (!args.Validate(unit))
             return 0;

**Comment by Aokromes** [comment]
Any feedback on https://github.com/TrinityCore/TrinityCore/issues/12817#issuecomment-345118237 ?

**Comment by Foereaper** [comment]
I'll test it as soon as I get my test server back up

**Comment by tkrokli** [comment]
Line numbers updated to match commit 46fcbb42d2ad

```diff
diff --git a/src/server/game/Movement/Spline/MoveSplineInit.cpp b/src/server/game/Movement/Spline/MoveSplineInit.cpp
index 02b31cfa19..997c0f7720 100644
--- a/src/server/game/Movement/Spline/MoveSplineInit.cpp
+++ b/src/server/game/Movement/Spline/MoveSplineInit.cpp
@@ -102,18 +102,13 @@ namespace Movement
         if (moveFlags & MOVEMENTFLAG_ROOT)
             moveFlags &= ~MOVEMENTFLAG_MASK_MOVING;
 
-        if (!args.HasVelocity)
-        {
-            // If spline is initialized with SetWalk method it only means we need to select
-            // walk move speed for it but not add walk flag to unit
-            uint32 moveFlagsForSpeed = moveFlags;
-            if (args.walk)
-                moveFlagsForSpeed |= MOVEMENTFLAG_WALKING;
-            else
-                moveFlagsForSpeed &= ~MOVEMENTFLAG_WALKING;
+        if (args.flags.walkmode)
+            moveFlags |= MOVEMENTFLAG_WALKING;
+        else
+            moveFlags &= ~MOVEMENTFLAG_WALKING;
 
-            args.velocity = unit->GetSpeed(SelectSpeedType(moveFlagsForSpeed));
-        }
+        if (!args.HasVelocity)
+            args.velocity = unit->GetSpeed(SelectSpeedType(moveFlags));
 
         if (!args.Validate(unit))
             return 0;
```

**Comment by tkrokli** [comment]
BTW, more change is needed to make that code compile, I get this error message (release):
```
C:\Trinity\Source\3.3.5\src\server\game\Movement\Spline\MoveSplineInit.cpp(105): error C2039: 'walkmode': is not a member of 'Movement::MoveSplineFlag'
c:\trinity\source\3.3.5\src\server\game\movement\spline\MoveSplineFlag.h(29): note: see declaration of 'Movement::MoveSplineFlag'
```
so we need to group more lines together with this excerpt/snippet.

**Comment by Nitair** [comment]
I can see that in first patch https://github.com/TrinityCore/TrinityCore/issues/12817#issuecomment-345118237 that `if (args.flags.walkmode)` got was previously here... but in https://github.com/TrinityCore/TrinityCore/issues/12817#issuecomment-357522188 `if (args.walk)` got replaced to `if (args.flags.walkmode)`. I guess at this point that error in https://github.com/TrinityCore/TrinityCore/issues/12817#issuecomment-357527861 is related to mixed different patches. So it should be changed to:

```
diff --git a/src/server/game/Movement/Spline/MoveSplineInit.cpp b/src/server/game/Movement/Spline/MoveSplineInit.cpp
index 02b31cfa19..997c0f7720 100644
--- a/src/server/game/Movement/Spline/MoveSplineInit.cpp
+++ b/src/server/game/Movement/Spline/MoveSplineInit.cpp
@@ -102,18 +102,13 @@ namespace Movement
         if (moveFlags & MOVEMENTFLAG_ROOT)
             moveFlags &= ~MOVEMENTFLAG_MASK_MOVING;
 
-        if (!args.HasVelocity)
-        {
-            // If spline is initialized with SetWalk method it only means we need to select
-            // walk move speed for it but not add walk flag to unit
-            uint32 moveFlagsForSpeed = moveFlags;
-            if (args.walk)
-                moveFlagsForSpeed |= MOVEMENTFLAG_WALKING;
-            else
-                moveFlagsForSpeed &= ~MOVEMENTFLAG_WALKING;
+        if (args.walk)
+            moveFlags |= MOVEMENTFLAG_WALKING;
+        else
+            moveFlags &= ~MOVEMENTFLAG_WALKING;
 
-            args.velocity = unit->GetSpeed(SelectSpeedType(moveFlagsForSpeed));
-        }
+        if (!args.HasVelocity)
+            args.velocity = unit->GetSpeed(SelectSpeedType(moveFlags));
 
         if (!args.Validate(unit))
             return 0;
```

**Comment by tkrokli** [comment]
Very well, thanks for the update. I will see if I can find time to test it during this week.
(all other TC 3.3.5 users, feel free to test it too) :-)

**Comment by CraftedRO** [comment]
I tested this on 80f30d5ca94f and for some reasons .go c id 32711 his pet remain stuck back in a dalaran room in his walk .

**Comment by illfated** [comment]
@CraftedRO : Would that be a specific room, or did it happen in any building visited by her (Warp-Huntress Kula) and her pet?

**Comment by CraftedRO** [comment]
Yes is about a specific chamber : The Agronomical Apothecary where is the alchemy trainer at 41.6 31.2 Waypoint 66 right on the exit corner of the door (where the flowerpot is).

**Comment by illfated** [comment]
Thanks, that is a useful tip for where it could happen. I will keep looking for that to happen on my game client (enGB).

**Comment by Aokromes** [comment]
if anything here is correct, plz make one PR.

**Comment by Legoso** [comment]
I was going to test the suggested diff in https://github.com/TrinityCore/TrinityCore/issues/12817#issuecomment-370153067 , but I was unable to adapt the diff code to the current code in the file (`src/server/game/Movement/Spline/MoveSplineInit.cpp`). Any chance that someone can post an updated diff that does not break the current code, so I may test it and provide feedback? (75cbafd or newer)
